walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,86 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
 
 public final class FindMeetingQuery {
+
+  /**
+  * Handles optional attendees and determines who to include in the meeting request
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+
+    // Return all times if there are no attendees
+    if (request.getAttendees().isEmpty() && request.getOptionalAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } 
+
+    ArrayList<String> mandatoryAttendees = new ArrayList<String>(request.getAttendees());
+    ArrayList<String> allAttendees = new ArrayList<String>(mandatoryAttendees);
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    ArrayList<TimeRange> timeRanges = new ArrayList<TimeRange>(findAvailableTimeRanges(allAttendees, events, duration));
+
+    // If no time ranges are available, exclude optional attendees and try again
+    if (!timeRanges.isEmpty()) {
+      return timeRanges;
+    } else {
+      timeRanges = new ArrayList<TimeRange>(findAvailableTimeRanges(mandatoryAttendees, events, duration));
+    }
+
+    // If the request has no attendees after optional attendees are excluded, return an empty collection.
+    if (!mandatoryAttendees.isEmpty()) {
+      return timeRanges;
+    } else {
+      return new ArrayList<TimeRange>();
+    }
+  }
+
+  /**
+  * Finds a collection of time ranges available for all of the specified attendees
+  */
+  public Collection<TimeRange> findAvailableTimeRanges(ArrayList<String> attendees, Collection<Event> events, long duration) {
+    ArrayList<TimeRange> timeRanges = new ArrayList<>();
+    ArrayList<TimeRange> badTimeRanges = new ArrayList<>();
+
+    // Make a list of all unavailable time ranges for all attendees
+    for (Event event: events) {
+      for (String attendee: attendees) {
+        if (event.getAttendees().contains(attendee)) {
+          badTimeRanges.add(event.getWhen());
+        }
+      }
+    }
+
+    if (!badTimeRanges.isEmpty()) {
```

Comment:
nit: we can remove this badTimeRanges.isEmpty(). Collections.sort no-ops if list to sort is empty

Reply:
done.